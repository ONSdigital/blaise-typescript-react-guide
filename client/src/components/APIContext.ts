import React from 'react';

interface APIContextValue {
  endpoint: string;
}

const APIContext = React.createContext<APIContextValue>({ endpoint: 'http://locahost:5000' });

export default APIContext;
