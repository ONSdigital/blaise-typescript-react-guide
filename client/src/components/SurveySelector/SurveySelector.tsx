import React, { ReactElement, ReactNode } from 'react';
import SurveyLoader from '../SurveyLoader';
import LoadingSpinner from '../LoadingSpinner';
import ErrorMessage from '../ErrorMessage/ErrorMessage';
import MultipleChoice from '../MultipleChoice';

interface SurveySelectorProps {
  onUpdate: (selected: string[]) => void
}

export default function SurveySelector({ onUpdate }: SurveySelectorProps): ReactElement {
  function errorMessage(error: string): ReactNode {
    return <ErrorMessage message={error} />;
  }

  return (
    <div>
      <SurveyLoader
        loadingMessage={<LoadingSpinner />}
        errorMessage={(error) => errorMessage(error)}
      >
        {(surveys) => <MultipleChoice question="Which surveys are your favourite?" values={surveys} onUpdate={onUpdate} />}
      </SurveyLoader>
    </div>
  );
}
