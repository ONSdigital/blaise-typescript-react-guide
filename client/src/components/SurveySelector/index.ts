import SurveySelector from './SurveySelector';

export default SurveySelector;
