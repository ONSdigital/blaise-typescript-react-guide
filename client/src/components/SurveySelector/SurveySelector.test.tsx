import React from 'react';
import {
  act, fireEvent, render, RenderResult, screen,
} from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import SurveySelector from './SurveySelector';
import APIContext from '../APIContext';

describe('SurveySelector', () => {
  const httpMock = new MockAdapter(axios);
  const apiContext = { endpoint: 'http://example.com/api' };

  let wrapper: RenderResult;
  let onUpdate: jest.Mock;

  function renderSurveySelector() {
    onUpdate = jest.fn();
    wrapper = render(
      <APIContext.Provider value={apiContext}>
        <SurveySelector onUpdate={onUpdate} />
      </APIContext.Provider>,
    );
  }

  describe('when loading', () => {
    beforeEach(() => {
      httpMock.onGet('http://example.com/api/surveys')
        .reply(() => new Promise(() => { /* never resolve */ }));
      renderSurveySelector();
    });

    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when surveys are listed', () => {
    beforeEach(() => {
      httpMock.onGet('http://example.com/api/surveys')
        .reply(200, ['SURVEY1', 'SURVEY2']);
      renderSurveySelector();
    });

    it('renders the select surveys message', async () => {
      expect(await screen.findByText('Which surveys are your favourite?')).toBeInTheDocument();
    });

    it('renders the list of surveys', async () => {
      expect(await screen.findByText('SURVEY1')).toBeInTheDocument();
      expect(await screen.findByText('SURVEY1')).toBeInTheDocument();
    });

    it('calls onUpdate when options are checked/unchecked', async () => {
      await screen.findByText('SURVEY1');
      act(() => {
        fireEvent.click(screen.getByLabelText('SURVEY1'));
        fireEvent.click(screen.getByLabelText('SURVEY2'));
        fireEvent.click(screen.getByLabelText('SURVEY1'));
      });
      expect(onUpdate).toHaveBeenCalledTimes(3);
      expect(onUpdate.mock.calls[0]).toContainEqual(['SURVEY1']);
      expect(onUpdate.mock.calls[1]).toContainEqual(['SURVEY1', 'SURVEY2']);
      expect(onUpdate.mock.calls[2]).toContainEqual(['SURVEY2']);
    });

    test('snapshot', async () => {
      await screen.findByText('SURVEY1');
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when a loading error occurs', () => {
    beforeEach(() => {
      httpMock.onGet('http://example.com/api/surveys')
        .reply(500, {});
      renderSurveySelector();
    });

    test('snapshot', async () => {
      await screen.findByText('Error:');
      expect(wrapper).toMatchSnapshot();
    });
  });
});
