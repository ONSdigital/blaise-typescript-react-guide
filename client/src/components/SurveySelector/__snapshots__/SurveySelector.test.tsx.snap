// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveySelector when a loading error occurs snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="survey-loader"
        >
          <div
            class="survey-loader__error"
          >
            <div
              class="error-message"
            >
              <div
                class="error-message__title"
              >
                Error:
              </div>
              <div
                class="error-message__message"
              >
                Error: Request failed with status code 500
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="survey-loader"
      >
        <div
          class="survey-loader__error"
        >
          <div
            class="error-message"
          >
            <div
              class="error-message__title"
            >
              Error:
            </div>
            <div
              class="error-message__message"
            >
              Error: Request failed with status code 500
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SurveySelector when loading snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="survey-loader"
        >
          <div
            class="survey-loader__loading"
          >
            <div
              class="loading-spinner"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="survey-loader"
      >
        <div
          class="survey-loader__loading"
        >
          <div
            class="loading-spinner"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SurveySelector when surveys are listed snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="survey-loader"
        >
          <div
            class="survey-loader__loaded"
          >
            <div
              class="multiple-choice"
            >
              <div
                class="multiple-choice__question"
              >
                Which surveys are your favourite?
              </div>
              <div
                class="multiple-choice__options"
              >
                <div
                  class="checkbox"
                >
                  <input
                    class="checkbox__input"
                    id=":r6:"
                    type="checkbox"
                  />
                  <label
                    class="checkbox__label"
                    for=":r6:"
                  >
                    SURVEY1
                  </label>
                </div>
                <div
                  class="checkbox"
                >
                  <input
                    class="checkbox__input"
                    id=":r7:"
                    type="checkbox"
                  />
                  <label
                    class="checkbox__label"
                    for=":r7:"
                  >
                    SURVEY2
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="survey-loader"
      >
        <div
          class="survey-loader__loaded"
        >
          <div
            class="multiple-choice"
          >
            <div
              class="multiple-choice__question"
            >
              Which surveys are your favourite?
            </div>
            <div
              class="multiple-choice__options"
            >
              <div
                class="checkbox"
              >
                <input
                  class="checkbox__input"
                  id=":r6:"
                  type="checkbox"
                />
                <label
                  class="checkbox__label"
                  for=":r6:"
                >
                  SURVEY1
                </label>
              </div>
              <div
                class="checkbox"
              >
                <input
                  class="checkbox__input"
                  id=":r7:"
                  type="checkbox"
                />
                <label
                  class="checkbox__label"
                  for=":r7:"
                >
                  SURVEY2
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
